File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn- yelling? [s]
  (let [s (str/replace s #"[^a-zA-Z]" "")]
    (and (not= "" s)(= s (str/upper-case s)))))

(defn- questioning? [s]
  (str/ends-with? (str/trimr s) "?"))

(defn- silence? [s]
  (str/blank? s))

(defn response-for [s]
  (let [yelling (yelling? s)
        questioning (questioning? s)
       ]
    (cond 
      (and yelling questioning) "Calm down, I know what I'm doing!" 
      yelling "Whoa, chill out!"
      questioning "Sure." 
      (silence? s) "Fine. Be that way!"
      :else "Whatever.")))

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where
import Data.List (all, dropWhile)
import Data.Char (isUpper, isSpace, isLetter)

trimr :: String -> String
trimr = reverse . dropWhile (\x -> isSpace x) . reverse

isQuestion :: String -> Bool
isQuestion p = let tp = trimr p in if null tp then False else last tp == '?'

isYell :: String -> Bool
isYell phrase = let letters = filter (\x -> isLetter x) phrase in not (null letters) && all (\x -> isUpper x) letters

isSilence :: String -> Bool
isSilence = null . filter (\x -> not $ isSpace x)

responseFor :: String -> String
responseFor xs
	| isQuestion xs = if isYell xs then "Calm down, I know what I'm doing!" else "Sure."
	| isYell xs = "Whoa, chill out!"
	| isSilence xs = "Fine. Be that way!"
	| otherwise = "Whatever."
